web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TraceContextCorrelation.java

```diff
@@ -0,0 +1,288 @@
+                return;
+            }
+
+
+            String traceId = request.getHeader(CORRELATION_HEADER_NAME);
+
+            Traceparent outGoingTraceParent = null;
+
+            if (traceId == null || traceId.isEmpty()) {
+                outGoingTraceParent = new Traceparent();
+
+                // represents the id of the current request.
+                requestTelemetry.setId(outGoingTraceParent.getTraceId() + "-" + outGoingTraceParent.getSpanId());
+
+                // represents the trace-id of this distributed trace
+                requestTelemetry.getContext().getOperation().setId(outGoingTraceParent.getTraceId());
+
+                // set parentId as null because this is the is the originating request
+                requestTelemetry.getContext().getOperation().setParentId(null);
+            } else {
+                Traceparent incomingTraceparent = Traceparent.fromString(traceId);
+
+                // create outgoing traceParent using the incoming header
+                outGoingTraceParent = new Traceparent(0, incomingTraceparent.getTraceId(),
+                    null, incomingTraceparent.getTraceFlags());
+
+                // set id of this request
+                requestTelemetry.setId(outGoingTraceParent.getTraceId() + "-" + outGoingTraceParent.getSpanId());
+
+                // represents the trace-id of this distributed trace
+                requestTelemetry.getContext().getOperation().setId(outGoingTraceParent.getTraceId());
+
+                // represents the parent-id of this request which is combination of traceId and incoming spanId
+                requestTelemetry.getContext().getOperation().setParentId(outGoingTraceParent.getTraceId() + "-" +
+                    incomingTraceparent.getSpanId());
+
+            }
+
+            String tracestate = request.getHeader(CORRELATION_COMPANION_HEADER_NAME);
+            Tracestate tracestateObject = Tracestate.fromString(tracestate);
+            Map<String, String> tracestatePropertiesMap = getPropertiesMap(tracestateObject);
+
+            // create outbound tracestate to be propagated to downstream calls
+            Tracestate outboundTracestate = createOutboundTracestate(tracestatePropertiesMap, getAppId());
+            ThreadContext.getRequestTelemetryContext().setTracestate(outboundTracestate);
+
+            // Let the callee know the caller's AppId
+            addTracestateInResponseHeader(response);
+
+        } catch (Exception e) {
+            InternalLogger.INSTANCE.trace("unable to perform correlation :%s", ExceptionUtils.
+                getStackTrace(e));
+        }
+    }
+
+    /**
+     * This adds the tracestate in response header so that the Callee can know what is the caller's AppId.
+     */
+    private static void addTracestateInResponseHeader(HttpServletResponse response) {
+
+        if (response.containsHeader(CORRELATION_COMPANION_HEADER_NAME)) {
```

Comment:
What's the rationale here? Are we opting not to overwrite if something else added the correlation header?

If that's the case, we may want to log something or it could be difficult to determine why correlation isn't working when this evals to true.

Reply:
That is correct. However looking here, I don't see any case when this would happen. As this is the only method which sets the Header in the response. I think it is safe to remove.